File: app/src/main/java/com/example/reminderapp/database/ReminderDao.java
package com.example.reminderapp.database;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

@Dao
public interface ReminderDao {
    @Insert
    long insert(ReminderEntity reminder);

    @Update
    void update(ReminderEntity reminder);

    @Delete
    void delete(ReminderEntity reminder);

    @Query("SELECT * FROM reminders WHERE userId = :userId ORDER BY date ASC")
    List<ReminderEntity> getRemindersForUser(int userId);

    @Query("SELECT * FROM reminders WHERE userId = :userId AND isPrivate = 1 ORDER BY date ASC")
    List<ReminderEntity> getPrivateRemindersForUser(int userId);

    @Query("SELECT * FROM reminders WHERE userId = :userId AND isPrivate = 0 ORDER BY date ASC")
    List<ReminderEntity> getPublicRemindersForUser(int userId);

    @Query("SELECT * FROM reminders WHERE userId = :userId AND isCompleted = 1 ORDER BY date DESC")
    List<ReminderEntity> getCompletedRemindersForUser(int userId);
    
    @Query("SELECT * FROM reminders WHERE userId = :userId AND isCompleted = 1 ORDER BY date DESC, time DESC LIMIT 5")
    List<ReminderEntity> getRecentCompletedReminders(int userId);

    @Query("SELECT * FROM reminders WHERE userId = :userId AND date = :date AND isPrivate = 0 ORDER BY time ASC")
    List<ReminderEntity> getPublicRemindersForDate(int userId, String date);

    @Query("SELECT * FROM reminders WHERE id = :id LIMIT 1")
    ReminderEntity getReminderById(int id);

    @Query("SELECT * FROM reminders WHERE isCompleted = 0")
    List<ReminderEntity> getAllUncompletedReminders();
}

File: app/src/main/java/com/example/reminderapp/database/ReminderDao.kt
package com.example.reminderapp.database

import androidx.room.*

@Dao
interface ReminderDao {
    @Insert
    fun insert(reminder: ReminderEntity): Long

    @Update
    fun update(reminder: ReminderEntity)

    @Delete
    fun delete(reminder: ReminderEntity)

    @Query("SELECT * FROM reminders WHERE userId = :userId ORDER BY date ASC")
    fun getRemindersForUser(userId: Int): List<ReminderEntity>

    @Query("SELECT * FROM reminders WHERE userId = :userId AND isPrivate = 1 ORDER BY date ASC")
    fun getPrivateRemindersForUser(userId: Int): List<ReminderEntity>

    @Query("SELECT * FROM reminders WHERE userId = :userId AND isPrivate = 0 ORDER BY date ASC")
    fun getPublicRemindersForUser(userId: Int): List<ReminderEntity>

    @Query("SELECT * FROM reminders WHERE userId = :userId AND isCompleted = 1 ORDER BY date DESC")
    fun getCompletedRemindersForUser(userId: Int): List<ReminderEntity>
    
    @Query("SELECT * FROM reminders WHERE userId = :userId AND isCompleted = 1 ORDER BY date DESC, time DESC LIMIT 5")
    fun getRecentCompletedReminders(userId: Int): List<ReminderEntity>

    @Query("SELECT * FROM reminders WHERE userId = :userId AND date = :date AND isPrivate = 0 ORDER BY time ASC")
    fun getPublicRemindersForDate(userId: Int, date: String): List<ReminderEntity>
}